src/AuthenticationService.php

```diff
@@ -317,4 +334,48 @@ public function buildIdentity($identityData)
 
         return $identity;
     }
+
+    /**
+     * Return the URL to redirect unauthenticated users to.
+     *
+     * If the `unauthenticaedRedirect` option is not set,
+     * this method will return null.
+     *
+     * If the `queryParam` option is set a query parameter
+     * will be appended with the denied URL path.
+     *
+     * @param Psr\Http\Message\ServerRequestInterface $request The request
+     * @return string|null
+     */
+    public function getUnauthenticatedRedirectUrl(ServerRequestInterface $request)
+    {
+        $param = $this->getConfig('queryParam');
+        $target = $this->getConfig('unauthenticatedRedirect');
+        if ($target === null) {
+            return null;
+        }
+        if ($param === null) {
+            return $target;
+        }
+
+        $uri = $request->getUri();
+        if (property_exists($uri, 'base')) {
```

Comment:
This makes me wonder if we should have our own Uri class/interface with additional methods for setting/getting base.

Reply:
Perhaps, we only use the base property in a few places. If it comes up more extending the `UriInterface` is a good idea.